### PR TITLE
Add validation tests for mapAsync.

### PR DESCRIPTION
**Pull request by Kangz**
PTAL!

### CODE REVIEW

File: src/webgpu/api/validation/buffer_mapping.spec.ts [hunk 1]
```diff
@@ -0,0 +1,763 @@
...
+    const kStride = 256;
+    const kNumStrides = 256;
+
+    const buffer = t.device.createBuffer({
+      size: kStride * kNumStrides,
+      usage: GPUBufferUsage.MAP_READ,
+    });
+    await buffer.mapAsync(GPUMapMode.READ);
+
+    // Get the a lot of smalle mapped ranges.
```
Comment (kainino0x) [inline src/webgpu/api/validation/buffer_mapping.spec.ts:641]:
```suggestion
    // Get a lot of small mapped ranges.
```

**Review by kainino0x — APPROVED** [review]

File: src/webgpu/api/validation/buffer_mapping.spec.ts [hunk 2]
```diff
@@ -0,0 +1,763 @@
...
+    with various cases at the limits of the buffer size or with a misaligned offset.
+    Also test for an empty buffer.`
+  )
+  .params(
+    params()
+      .combine(kMapModeOptions)
+      .combine([
+        // 0 size buffer.
+        { bufferSize: 0, offset: 0 },
+        // Check that the defaulting of size is done before checking the alignment for offset.
```
[the remark this answers is not in the document]
Comment (Kangz) [reply src/webgpu/api/validation/buffer_mapping.spec.ts:206]: Ah, it is no longer distinguishable. Removed the comment.

File: src/webgpu/api/validation/buffer_mapping.spec.ts [hunk 3]
```diff
@@ -0,0 +1,763 @@
...
+        unreachable();
+    }
+  }
+}
+
+export const g = makeTestGroup(F);
+
+const kMapModeOptions = poptions('mapMode', [GPUMapMode.READ, GPUMapMode.WRITE]);
+const kOffsetAlignment = 8;
+const kSizeAlignement = 4;
```
[the remark this answers is not in the document]
Comment (Kangz) [reply src/webgpu/api/validation/buffer_mapping.spec.ts:75]: Whoops frenchism.

File: src/webgpu/api/validation/buffer_mapping.spec.ts [hunk 4]
```diff
@@ -0,0 +1,763 @@
...
+    const buffer = t.createMappableBuffer(mapMode, bufferSize);
+    await buffer.mapAsync(mapMode, mapOffset, mapSize);
+
+    const actualMapOffset = mapOffset ? mapOffset : 0;
+    const actualMapSize = mapSize ? mapSize : bufferSize - actualMapOffset;
+
+    const actualOffset = offset ? offset : 0;
+
+    let success = actualOffset >= actualMapOffset && actualOffset <= bufferSize;
+    if (size !== undefined) {
```
[the remark this answers is not in the document]
Comment (Kangz) [reply src/webgpu/api/validation/buffer_mapping.spec.ts:569]: Fixed the condition. The tests weren't failing when success was true but `getMappedRange` failed because the test function was async and the exception was thrown in the returned promise.

File: src/webgpu/api/validation/buffer_mapping.spec.ts [hunk 5]
```diff
@@ -0,0 +1,763 @@
...
+
+g.test('getMappedRange_state_mappingPending')
+  .desc('Test that it is invalid to call getMappedRange in the mappingPending state.')
+  .params(kMapModeOptions)
+  .fn(t => {
+    const { mapMode } = t.params;
+    const buffer = t.createMappableBuffer(mapMode, 16);
+
+    buffer.mapAsync(mapMode);
+    t.testGetMappedRangeCall(false, buffer);
```
[the remark this answers is not in the document]
Comment (Kangz) [reply src/webgpu/api/validation/buffer_mapping.spec.ts:393]: Yeah I think these are more operation tests. We'd do timeline checks there.

File: src/webgpu/api/validation/buffer_mapping.spec.ts [hunk 6]
```diff
@@ -0,0 +1,763 @@
...
+  });
+
+g.test('getMappedRange_disjointRanges')
+  .desc('Test that the ranges asked through getMappedRange must be disjoint.')
+  .params(
+    params()
+      .combine(pbool('remapBetweenCalls'))
+      .combine([
+        // Disjoint ranges with one that's empty.
+        { offset1: 8, size1: 0, offset2: 8, size2: 8 },
```
[the remark this answers is not in the document]
Comment (Kangz) [reply src/webgpu/api/validation/buffer_mapping.spec.ts:583]: Wouldn't that make the test more verbose because `offset1` would need to become `ranges[0].offset`? Can typescript don't seem to be able to destruct-assign sequences of dict?

I was thinking of focusing this test on only two ranges so that the logic can stay somewhat simple (and not do loops of loops)

File: src/webgpu/api/validation/buffer_mapping.spec.ts [hunk 7]
```diff
@@ -0,0 +1,763 @@
...
+
+g.test('mapAsync_state_destroyed')
+  .desc('Test that mapAsync is an error when called on a destroyed buffer.')
+  .params(kMapModeOptions)
+  .fn(async t => {
+    const { mapMode } = t.params;
+    const buffer = t.createMappableBuffer(mapMode, 16);
+    buffer.destroy();
+    await t.testMapAsyncCall(false, 'OperationError', buffer, mapMode);
+  });
```
[the remark this answers is not in the document]
Comment (Kangz) [reply src/webgpu/api/validation/buffer_mapping.spec.ts:125]: I think it makes sense for other tests because the usage is only used server-side, but mapping usage could be checked client-side so we need to make sure the usage flags are correct (which a helper wouldn't give us). WDYT?

File: src/webgpu/api/validation/buffer_mapping.spec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (kainino0x) [reply src/webgpu/api/validation/buffer_mapping.spec.ts:590]: You can nest destructurings `const { ranges: [{ offset: offset1, size: size1 }, { offset: offset2, size: size2 }] } = t.params;`
but I was thinking about doing it with a loop. That would be overly complex if you don't want to do more than 2 ranges.

File: src/webgpu/api/validation/buffer_mapping.spec.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (kainino0x) [reply src/webgpu/api/validation/buffer_mapping.spec.ts]: createBufferWithState does take a descriptor. It ignores it if you ask for an error texture, but it doesn't necessarily have to.

**Comment by kainino0x** [comment]
I'll be out tomorrow, feel free to merge when ready. (I'd probably just merge it now but it's failing CI)

**Review by kainino0x — APPROVED** [review]

File: src/webgpu/api/validation/buffer_mapping.spec.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Kangz) [reply src/webgpu/api/validation/buffer_mapping.spec.ts]: I had some trouble refactoring to use the new helper because all tests are parametrized with the `mapAsync` mode and not the buffer usage so the helper requires extra steps to use (which kinda defeat the purpose).

I still included it in in the latest patchset.
